### PR TITLE
Fixes #29507 - Add puma-status for reporting

### PR DESCRIPTION
**Pull request by sthirugn**
Note the foreman packaging change is still pending to include `puma-status` gem as rpm

### CODE REVIEW

**Comment by sthirugn** [comment]
this PR requires packaging of https://rubygems.org/gems/puma-status/versions/1.0 in https://github.com/theforeman/foreman-packaging

File: config/puma/production.rb
Comment (sthirugn) [inline config/puma/production.rb]: I chose port 9293 as per the [puma config recommendation](https://github.com/puma/puma#controlstatus-server). Do we have any restrictions on choosing the port numbers in foreman?

File: bundler.d/service.rb
Comment (sthirugn) [inline bundler.d/service.rb]: https://rubygems.org/gems/puma-status/versions/1.0

File: config/puma/production.rb
Comment (timogoebel) [reply config/puma/production.rb]: Users should be able to change the port. The best approach would be to use environment variables.
The `auth_token` needs to be secure. A default of `12345` is unacceptable if it's used for authentication.

Comment (ekohl) [reply config/puma/production.rb]: Can we listen on a unix socket perhaps? That way we can use file permissions to limit access.

File: script/foreman-puma-status
Comment (ekohl) [inline script/foreman-puma-status]: We've always kept SCL out of the main Foreman repository because it's only for EL7. It doesn't make sense in Debian, Ubuntu and EL8. This is something we apply in packaging.

I also wonder how this path is chosen. Foreman has a temp dir and perhaps that makes more sense? Or `/run/foreman/puma.state`

File: bundler.d/service.rb
Comment (ekohl) [reply bundler.d/service.rb]: This will need a PR to https://github.com/theforeman/foreman-packaging/tree/rpm/develop to introduce `puma-status`.

File: config/puma/production.rb
Comment (lzap) [reply config/puma/production.rb]: Whatever we do, let me know and I will create a change in SELinux policy. I prefer local socket to be honest, there is no need to expose this over TCP. I'd ditch token in this case since only puma/foreman/root users should be able to read it:

    activate_control_app 'unix:///var/run/foreman-pumactl.sock'

Oh it actually exposes also actions like start/stop/restart. We might consider tokens as well, could be stored in `/etc/foreman/settings.yml` and read by the CLI app. But that's security by obscurity rather than something good.

File: bundler.d/service.rb
Comment (lzap) [reply bundler.d/service.rb]: Isn't `puma-status` a CLI app rather than server app? This does not need to be here at all.

Comment (sthirugn) [reply bundler.d/service.rb]: > This will need a PR to https://github.com/theforeman/foreman-packaging/tree/rpm/develop to introduce `puma-status`.

I asked  Zach to add puma-status to foreman-packaging.

Comment (sthirugn) [reply bundler.d/service.rb]: > Isn't `puma-status` a CLI app rather than server app? This does not need to be here at all.

`puma-status` is a separate gem and just a cli app. Are you saying we can add this to `foreman-packaging` and remove this entry from here?

File: script/foreman-puma-status
Comment (sthirugn) [reply script/foreman-puma-status]: > We've always kept SCL out of the main Foreman repository because it's only for EL7. It doesn't make sense in Debian, Ubuntu and EL8. This is something we apply in packaging.

For my understanding, if I just keep `puma-status /tmp/puma/pids/puma.state`, how does it work?

Comment (sthirugn) [reply script/foreman-puma-status]: > I also wonder how this path is chosen. Foreman has a temp dir and perhaps that makes more sense? Or `/run/foreman/puma.state`
 
I missed pushing a change yesterday, I intended to use `/usr/share/foreman/tmp/pids/` since we already have it created for other process ids. 

I updated it to use `/var/run/foreman`

File: config/puma/production.rb
Comment (sthirugn) [inline config/puma/production.rb]: puma automatically creates auth token, so we don't need to explicitly supply it.

File: config/puma/production.rb
Comment (sthirugn) [reply config/puma/production.rb]: Yes the sockets are better.  Now I added
`activate_control_app 'unix:///var/run/foreman/foreman-pumactl.sock'`
and puma automatically created the auth token for me.
```console
# cat /var/run/foreman/pids/puma.state 
---
pid: 7200
control_url: unix:///var/run/foreman/foreman-pumactl.sock
control_auth_token: 70443a6f423955719497c4c91df4ff
```

**Comment by sthirugn** [comment]
I squashed all the commits, test and the output is:

```console
# systemctl restart foreman

# journalctl -u foreman -f
Apr 09 18:57:39 centos7-katello-nightly.apex.example.com systemd[1]: Started Foreman.
Apr 09 18:57:42 centos7-katello-nightly.apex.example.com foreman[7802]: /usr/share/foreman/lib/foreman.rb:8: warning: already initialized constant Foreman::UUID_REGEXP
Apr 09 18:57:42 centos7-katello-nightly.apex.example.com foreman[7802]: /usr/share/foreman/lib/foreman.rb:8: warning: previous definition of UUID_REGEXP was here
Apr 09 18:57:43 centos7-katello-nightly.apex.example.com foreman[7802]: => Booting Puma
Apr 09 18:57:43 centos7-katello-nightly.apex.example.com foreman[7802]: => Rails 5.2.1 application starting in production
Apr 09 18:57:43 centos7-katello-nightly.apex.example.com foreman[7802]: => Run `rails server -h` for more startup options
Apr 09 18:57:55 centos7-katello-nightly.apex.example.com foreman[7802]: [7802] Puma starting in cluster mode...
Apr 09 18:57:55 centos7-katello-nightly.apex.example.com foreman[7802]: [7802] * Version 4.3.3 (ruby 2.5.5-p157), codename: Mysterious Traveller
Apr 09 18:57:55 centos7-katello-nightly.apex.example.com foreman[7802]: [7802] * Min threads: 0, max threads: 16
Apr 09 18:57:55 centos7-katello-nightly.apex.example.com foreman[7802]: [7802] * Environment: production
Apr 09 18:57:55 centos7-katello-nightly.apex.example.com foreman[7802]: [7802] * Process workers: 2
Apr 09 18:57:55 centos7-katello-nightly.apex.example.com foreman[7802]: [7802] * Phased restart available
Apr 09 18:57:55 centos7-katello-nightly.apex.example.com foreman[7802]: [7802] * Activated tcp://127.0.0.1:3000
Apr 09 18:57:55 centos7-katello-nightly.apex.example.com foreman[7802]: [7802] Use Ctrl-C to stop
Apr 09 18:57:55 centos7-katello-nightly.apex.example.com foreman[7802]: [7802] * Starting control server on unix:///var/run/foreman/foreman-pumactl.sock
Apr 09 18:57:55 centos7-katello-nightly.apex.example.com foreman[7802]: [7802] - Worker 0 (pid: 7842) booted, phase: 0
Apr 09 18:57:55 centos7-katello-nightly.apex.example.com foreman[7802]: [7802] - Worker 1 (pid: 7846) booted, phase: 0

# ll /var/run/foreman/pids
total 12
-rw-r--r--. 1 foreman foreman   5 Apr  9 18:57 puma.pid
-rw-r--r--. 1 foreman foreman 124 Apr  9 18:57 puma.state
-rw-r--r--. 1 foreman foreman   4 Apr  9 18:57 server.pid

# cat /var/run/foreman/pids/puma.state
---
pid: 7802
control_url: unix:///var/run/foreman/foreman-pumactl.sock
control_auth_token: f498f3f689f023ee6afe6e82699f36c

# cat /var/run/foreman/pids/puma.pid
7802

# /bin/bash /usr/share/foreman/script/foreman-puma-status 
7802 (/var/run/foreman/pids/puma.state) Uptime:  2m17s | Phase: 0 | Load: 0[░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░]32
 └  7842 CPU:   0.0% Mem:  294 MB Uptime:  2m17s | Load: 0[░░░░░░░░░░░░░░░░]16
 └  7846 CPU:   0.0% Mem:  294 MB Uptime:  2m17s | Load: 0[░░░░░░░░░░░░░░░░]16
```

File: script/foreman-puma-status
Comment (ehelms) [inline script/foreman-puma-status]: No `scl enable` needed here. We treat this like a standard bash script so it works everywhere. Then our packaging environments can adjust this as needed. For example, in EL7 environments, we update this to change the entrypoint:

```
#!/bin/bash
```

becomes

```
#!/usr/bin/tfm-ruby
```

File: bundler.d/service.rb
Comment (ehelms) [reply bundler.d/service.rb]: That makes sense. If it's ready the state file and not actually hooked into the running server then it won't need to be declared here and we can make it a dependency of the `foreman-service` package.

Comment (sthirugn) [reply bundler.d/service.rb]: thanks, removed `puma-status` from `bundler.d/service.rb`

File: script/foreman-puma-status
Comment (sthirugn) [reply script/foreman-puma-status]: ok I removed `scl` and made it like a simple bash script.

File: config/puma/production.rb
Comment (ekohl) [inline config/puma/production.rb]:
```suggestion
run_dir = Rails.root.join('tmp')

# Store the pid of the server.
pidfile File.join(run_dir, 'pids', 'puma.pid')

# Store server info state.
state_path File.join(run_dir, 'puma.state')
```

File: config/puma/production.rb
Comment (ekohl) [reply config/puma/production.rb]:
```suggestion
activate_control_app "unix://#{run_dir}/sockets/pumactl.sock"
```

File: script/foreman-puma-status
Comment (ekohl) [inline script/foreman-puma-status]:
```suggestion
puma-status /run/foreman/puma.state
```

**Review by ekohl — COMMENTED** [review]
The use of `/var/run` is generally deprecated since systemd introduced `/run`. It is also a symlink on all platforms we support now. In packaging we ensure this directory exists by creating `/usr/lib/tmpfiles.d/foreman.conf`.

On a related, note, we still symlink `/usr/share/foreman/tmp` to `/var/run/foreman` instead of `/run/foreman`. A brief look tells me we should be using `RuntimeDirectory=foreman` in `foreman.service` but then we also need to take a look at the `cache` directory. The more I looked into this, the more things I saw to improve. https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RuntimeDirectory= has a nice list but Rails doesn't have a variable to change the `tmp` directory and just uses `Rails.root.join('tmp')` all over the place. That's why packaging symlinks it.

I also see that `/run/foreman/sockets` exists, but is empty. It appears this is a Rails convention and we should be making use of that.

You're now only configuring this for production, but shouldn't this be a general config? development also uses puma and it'd be nice to test the same tools there.

File: config/puma/production.rb
Comment (ekohl) [inline config/puma/production.rb]: Reading further about this, the default pid is `tmp/pids/server.pid` and I don't see the need to change this. I also think that `rails server` has a `--pid` option to change this and there's no real reason to hardcode it. In systemd we don't even care about pids at all since it's all managed for us. Am I missing something?

Comment (sthirugn) [reply config/puma/production.rb]: you are right, we are concerned about the `state_path` option only and there is no need to change `pidfile` option. I changed it to make sure the `pid` and `state` files live together, but it doesn't matter. I will remove this entry.

File: config/puma/production.rb
Comment (sthirugn) [reply config/puma/production.rb]: thanks @ekohl , to make it even simpler, should we perhaps just use 
`run_dir = "/run/foreman"`

**Comment by sthirugn** [comment]
> You're now only configuring this for production, but shouldn't this be a general config? development also uses puma and it'd be nice to test the same tools there.

I can submit a change for dev environment, which repository should this change be submitted to?

File: config/puma/production.rb
Comment (ekohl) [inline config/puma/production.rb]: I think this currently breaks running a development checkout when using `RAILS_ENV=production`. In our packaged setup we symlink `tmp` to this directory so you can make a compatible change:
```suggestion
run_dir = Rails.root.join('tmp')
```

Comment (sthirugn) [reply config/puma/production.rb:1]: To understand this better, can you explain the benefit of this change?

**Comment by sthirugn** [comment]
Squashed commits and ready for final review/merge.

**Comment by sthirugn** [comment]
katello failed tests don't seem to be related to this change.

**Review by ehelms — DISMISSED** [review]

File: script/foreman-puma-status
Comment (timogoebel) [inline script/foreman-puma-status:3]: Does this work on a Mac?

Comment (ekohl) [reply script/foreman-puma-status:3]: I doubt it does, but running `puma-status ~foreman/tmp/puma.state` probably does. Unless you use a different user in which case it doesn't really scale. You could accept an argument but then you may also need to use bundle exec. I think this file is more intended for production.

**Comment by ekohl** [comment]
Are there packaging PRs (both deb and RPM) to make sure `foreman-puma-status` is installed to the correct location?

File: script/foreman-puma-status
Comment (sthirugn) [reply script/foreman-puma-status:3]: Is the foreman server expected to be running on a Mac OS?

**Comment by sthirugn** [comment]
> Are there packaging PRs (both deb and RPM) to make sure `foreman-puma-status` is installed to the correct location?

I see this PR https://github.com/theforeman/foreman-packaging/pull/5031 manipulates `foreman-puma-status`. What other changes are needed?

File: script/foreman-puma-status
Comment (timogoebel) [reply script/foreman-puma-status:3]: Just the dev environment. I'm good with merging this, we can just add a comment `Intended for a production environment`.

Comment (sthirugn) [reply script/foreman-puma-status:3]: @timogoebel ty for the clarification. Added a comment in the script now.

**Comment by ehelms** [comment]
Packaging change for RPM:

https://github.com/theforeman/foreman-packaging/pull/5031

**Comment by ehelms** [comment]
Any further comments or concerns with this? With packaging PRs open, I believe this is GTG.

**Review by ehelms — DISMISSED** [review]

File: config/puma/production.rb
Comment (ekohl) [inline config/puma/production.rb:28]: I haven't checked the permissions on this, but I'm assuming it's `foreman:foreman` and `0644` which should prevent from everyone writing to it. Correct?

**Review by ekohl — DISMISSED** [review]
Approving if my assumption is correct.

File: config/puma/production.rb
Comment (sthirugn) [reply config/puma/production.rb:28]: This is how it looks now:
```console
# ll -d /run/foreman/ /run/foreman/sockets
drwxr-xr-x. 5 foreman foreman 200 Apr 22 14:30 /run/foreman/
drwxr-xr-x. 2 foreman foreman  60 Apr 22 14:23 /run/foreman/sockets

# ll /run/foreman/sockets
total 0
srwxrwxrwx. 1 foreman foreman 0 Apr 22 14:23 pumactl.sock
```

Comment (ekohl) [reply config/puma/production.rb:28]: Does that mean that any user can connect to the socket and restart the server?

Comment (ehelms) [reply config/puma/production.rb:28]: Are these determined by packaging?

Comment (sthirugn) [reply config/puma/production.rb:28]: yes, looks like a `postgres` user is able to restart the puma server if right `control_url` and `control_auth_token` is known. Any suggestions on how to lock this down further?
```console
# ll /var/run/foreman/puma.state 
-rw-r--r--. 1 foreman foreman 120 Apr 22 15:23 /var/run/foreman/puma.state

# cat /var/run/foreman/puma.state 
---
pid: 1672
control_url: unix:///run/foreman/sockets/pumactl.sock
control_auth_token: 6c445af0312fff70fe0f190a65644cf

# sudo -u postgres scl enable tfm 'pumactl --control-url 'unix:///var/run/foreman/sockets/pumactl.sock' --control-token 6c445af0312fff70fe0f190a65644cf  restart'
Command restart sent success
```

Comment (ekohl) [reply config/puma/production.rb:28]: https://github.com/puma/puma/blob/2668597ec1dd9546d83db9f2ec5ad092add483e6/lib/puma/state_file.rb#L11-L13 suggests it can't be configured to set permissions. Right now I don't have advise.

Comment (ehelms) [reply config/puma/production.rb:28]: What are the permissions on viewing the control token?

Comment (sthirugn) [reply config/puma/production.rb:28]: see the first line of my post above. the control token is present in puma.state file which has permissions 644

Comment (ehelms) [reply config/puma/production.rb:28]: All I can think is having something to adjust the permissions after creation but really the answer is to get a change into Puma to allow configuring the state file permissions.

Is it possible to put the state file in a directory that is not globally readable?

Comment (sthirugn) [reply config/puma/production.rb:28]: > All I can think is having something to adjust the permissions after creation but really the answer is to get a change into Puma to allow configuring the state file permissions.

I am working to get a fix to the puma project.

Comment (sthirugn) [reply config/puma/production.rb:28]: Submitted patch to puma project https://github.com/puma/puma/pull/2238 to address the state file security issue.

**Comment by tbrisker** [comment]
@sthirugn @ekohl What's the status here?

**Comment by ekohl** [comment]
Blocked on the linked PR (https://github.com/puma/puma/pull/2238). There's good progress on it.

**Comment by sthirugn** [comment]
Note that https://github.com/puma/puma/pull/2238 is merged now. So I added a new commit to this PR to add state_permission of `0644`

**Comment by ekohl** [comment]
> Note that [puma/puma#2238](https://github.com/puma/puma/pull/2238) is merged now. So I added a new commit to this PR to add state_permission of `0644`

Note that merged doesn't mean available. You'll need to wait for a gem to include it and require at least puma x.y.z.

**Comment by ehelms** [comment]
The related Puma fix is merged and released in Puma.

**Review by ekohl — CHANGES_REQUESTED** [review]
This should now require the right puma version but other than that :+1:

**Comment by tbrisker** [comment]
closing in favor of gh-7735

**Pull request closed without merging**
